clues.go

```diff
@@ -224,3 +224,63 @@ func AddLabelCounterTo(
 
 	return setNodeInCtx(ctx, namespace, nn)
 }
+
+// ---------------------------------------------------------------------------
+// comments
+// ---------------------------------------------------------------------------
+
+// AddProxy attaches up a new clues proxy in the context.  All proxies are
+// passed down to all descendants which branch off of this context.  Whenever
+// a clues is added to the context, it gets added to the proxy as well.  As
+// a result, any clues added to the context in descendent funcs will appear in
+// this context as well.
+//
+// If the proxyID is an empty string, a random id will be generated.
+//
+// If isolateProxyValues is true, the proxy will namespace all of its values
+// using the proxyID to avoid clobbering any existing values.
+//
+// Proxy support is specifically useful in a unique situation: when you are
+// able to both pass in a ctx and also able to add clues data to it, but also
+// unable to later retrieve the ctx or to bind the ctx values into an error.
+// This case can manifest when passing ad-hoc functions or middleware to
+// callers who limit options to control for error returns.
+func AddProxy(
+	ctx context.Context,
+	proxyID string,
+	isolateProxyValues bool,
+) context.Context {
+	nc := nodeFromCtx(ctx, defaultNamespace)
+	nn := nc.addValues(nil)
+	nn = nn.addProxy(proxyID, isolateProxyValues)
+
+	return setDefaultNodeInCtx(ctx, nn)
+}
+
+// AddProxyTo attaches up a new clues proxy within the context namespace.
+// All proxies are passed down to all descendants which branch off of this
+// context.  Whenever a clues is added to the context, it gets added to the
+// proxy as well.  As a result, any clues added to the context in descendent
+// funcs will appear in this context as well.
+//
+// If the proxyID is an empty string, a random id will be generated.
+//
+// If isolateProxyValues is true, the proxy will namespace all of its values
+// using the proxyID to avoid clobbering any existing values.
+//
+// Proxy support is specifically useful in a unique situation: when you are
+// able to both pass in a ctx and also able to add clues data to it, but also
+// unable to later retrieve the ctx or to bind the ctx values into an error.
+// This case can manifest when passing ad-hoc functions or middleware to
+// callers who limit options to control for error returns.
+func AddProxyTo(
```

Comment:
placing comment here because I dislike GitHub's lack of threading for comments on the review itself...

I think in general the implementation looks good, though I'm on the fence about the feature. My hesitancy comes from two main things:
1. there's no easy way to describe how proxied values will appear in the output (as in what prefixes they'll get, how overrides work with multiple children, etc)
2. there's no way to remove proxied data from a node

I think that this could put us in a position where the logs we need to read through contain too much information, and thus become less useful. We've gotten good at adding a lot of data to the context, but that's going to come back to hurt us here because we have no way to control what data is proxied, and to where. In essence, this is hoping that called functions know exactly what the caller is doing and when it's ok to add extra info and when it's not. Without that info, it will become easy to just assume it's safe to proxy things and end up overwhelmed by the amount of data we need to dig through. We may also start to clobber data we actually want to keep because a called function proxied something that just happened to have the same key as something in the parent (e.x. `id` is pretty common to use for keys)

Reply:
> because I dislike GitHub's lack of threading for comments on the review itself

Yo, tell me about it.  Why can't we just have threads on the main review?  Makes no sense.

> there's no way to remove proxied data from a node

permanently?  Or as in, you'd like to be able to fetch the values without including the proxies.

> In essence, this is hoping that called functions know exactly what the caller is doing and when it's ok to add extra info and when it's not.

Ah, the goal was to make that concern not appear.  Is there some inherent fuzziness to the data addition that isn't present in other ctx additions?  Yeah, unfortunately there is.  But this is also designed for an extremely weird corner case situation.  On which note...

> it will become easy to just assume it's safe to proxy things and end up overwhelmed by the amount of data we need to dig through

I am, honesty, riding a bit on the expectation that Proxies are almost never useful, and when they are useful it's for a very small and encapsulated set of data.  I do share the concern that this can blow out metadata in unwanted ways, especially if used haphazardly.  But there's _so little_ reason to use it, that it may balance out in the end?  Fingers crossed?

> We may also start to clobber data we actually want to keep because a called function proxied something that just happened to have the same key as something in the parent (e.x. id is pretty common to use for keys)

That was why proxies get namespaces by default.  In fact, the option to un-namespace a proxy was an offbeat afterthought.  Maybe that's unnecessary and a footgun more than it is helpful.

Your response is a bit balanced between two worries: too much data, and the accidental data clobbering. I think the latter is relatively a non-issue due to proxy value isolation (I think I'll do away with non-isolated variations).  The former is, well, like stated, based on minimal usage as a strategy.

Here's a pitch, what if this was more of a producer:consumer pairing.   I could introduce a func like `clues.AddToProxy(...)` that adds the value to both the ctx and (if one exists) the proxy handed to the context.  It feels overbearing.  But also feels like it could solve your concerns.  Thoughts?

Reply:
>> there's no way to remove proxied data from a node
>
> permanently? Or as in, you'd like to be able to fetch the values without including the proxies.

yea, permanently. For example, if you add a proxy to something so web requests can return enough info in logs but the context the proxy was added to is also used for a whole bunch of other things after making the web requests there's no way to remove the proxied info for the "whole bunch of other things" after the part where you actually wanted the proxy. I suppose you could stop using the context that has the proxy at some point, but then the question becomes "why is there one context for half of this function and a different context for the other half?"

>> it will become easy to just assume it's safe to proxy things and end up overwhelmed by the amount of data we need to dig through
>
> I am, honesty, riding a bit on the expectation that Proxies are almost never useful, and when they are useful it's for a very small and encapsulated set of data. I do share the concern that this can blow out metadata in unwanted ways, especially if used haphazardly. But there's so little reason to use it, that it may balance out in the end? Fingers crossed?

While I can understand the hope, it may not play out that way. This is something that won't have a correctness or (hopefully) perf impact, so it's easy to misuse. The fact that it's difficult to explain the full behavior of, not intended to be used much, and is difficult to track through the codebase (i.e. called functions must know their parents added a proxy and must know to be careful about how many clues they add) makes it likely it will be misused

> That was why proxies get namespaces by default. In fact, the option to un-namespace a proxy was an offbeat afterthought. Maybe that's unnecessary and a footgun more than it is helpful.

It's still possible for two proxied values to clobber each other, even with the namespacing

> I could introduce a func like clues.AddToProxy(...) that adds the value to both the ctx and (if one exists) the proxy handed to the context

This could help alleviate some of the concerns, but there's still the chance of clobbering proxied values if the same key is used. For example, a call chain of 3 functions, all of which add proxied values with the same key

Reply:
> This could help alleviate some of the concerns, but there's still the chance of clobbering proxied values if the same key is used. For example, a call chain of 3 functions, all of which add proxied values with the same key

I think I'm going to refactor with this design in mind.  If we come back with needing a broad scale proxy bucket, we can come back to this design again.

As for clobbering: last-value-wins is part and parcel of the package.  Not much way around it, without having a system of user-facing checks and balances, which is definitely too heavy.  In the case of proxies, I suppose I could ask callers to key their proxy on both the initialization and the data addition.  That would provide strong guarantees about isolation.  Jives well with producer-consumer relationships, too.